Add a caching OIDC validator for use with Kubernetes OIDC

timothyb89 opened the pull request:
This adds a generic caching OIDC validator which is initially used to cache Kubernetes OIDC requests. This takes a simpler approach to the one explored in #57789 and explicitly caches the two resources in question:
- We cache the OIDC discovery configuration for an hour
- We cache the JWKS keyset for 24 hours, and additionally allow the library to cache internally. It currently has no internal method for invalidating old keys, so we purge the `oidc.KeySet` regularly. However, upon encountering an unknown `kid`, the library will attempt to fetch new keys on its own.
- At most once an hour when getting a validator, we prune validators that have not been used in 48 hours, to ensure they don't remain in memory forever if e.g. new join tokens are added and removed over time.

This validator is designed to be generic and should be easy to apply to other users of our existing OIDC helper.

This isn't the smartest or most efficient caching implementation - I think I could spend a couple months rigorously writing and validating a cache - but given the load and concurrency we realistically expect, I think this is a simple enough implementation to get the job done without introducing too many new points of failure.

closes https://github.com/gravitational/teleport/issues/49213

changelog: Kubernetes OIDC responses are now cached to improve performance and reliability when joining bots and nodes

timothyb89 committed:
Add a caching OIDC validator for use with Kubernetes OIDC

This adds a generic caching OIDC validator which is initially used
to cache Kubernetes OIDC requests. This takes a simpler approach to
the one explored in #57789 and explicitly caches the two resources
in question:
- We cache the OIDC discovery configuration for an hour
- We cache the JWKS keyset for 24 hours, and additionally allow the
  library to cache internally. It currently has no internal method
  for invalidating old keys, so we purge the `oidc.KeySet` regularly.
  However, upon encountering an unknown `kid`, the library will
  attempt to fetch new keys on its own.

This validator is designed to be generic and should be easy to
apply to other users of our existing OIDC helper.

timothyb89 committed:
Manage validator instances by issuer; use in Kubernetes validator

This adds an indirection layer to automatically manage validator
instances as cached data is specific to a particular (issuer,
audience) combination.

Additionally, this swaps the new caching validator in in place of the
standard non-caching validator.

timothyb89 committed:
Fix comments and add missing mutex lock

timothyb89 committed:
Add tests for the caching validator

timothyb89 committed:
Fix lints

timothyb89 committed:
Add logging

timothyb89 committed:
Fix imports

timothyb89 committed:
Merge branch 'timothyb89/kubernetes-oidc-caching-validator' of github.com:gravitational/teleport into timothyb89/kubernetes-oidc-caching-validator

boxofrad reviewed (approved):
🚀

strideynet reviewed (approved):
Looks good to me - I don't have too many concerns about cache size running away since audience/issuer should always come from configuration in Teleport rather than the token itself. However, it may still be valuable to have someone from performance/scaling look this over.

strideynet commented:
Hey @timothyb89 - can we merge this down ?

timothyb89 committed:
Code review suggestions: atomic expiry, misc fixes

This changes `validatorExpires` to be an atomic int64, which removes
a locks from `Expires()` / `IsStale()` and ensures an ongoing pruning
won't block `ValidateToken()`.

timothyb89 commented on lib/oidc/caching_token_validator.go:
```diff
@@ -0,0 +1,307 @@
+
+	retained := map[validatorKey]*CachingValidatorInstance[C]{}
+	prunedCount := 0
+	for k, v := range v.instances {
```
I... don't know why I did that, apparently some part of my lizard brain was trying to treat it like a list instead of a map, d'oh. It just deletes in place, now.

timothyb89 committed:
Replace validator cache and pruning with `utils.FnCache`

This swaps the hand-rolled cache for validators with utils.FnCache,
since it's more robust and its TTL mechanism works properly for our
needs.

timothyb89 committed:
Remove commented out test code

timothyb89 committed:
Fix build errors and enable ReloadOnErr

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/kubernetes-oidc-caching-validator

timothyb89 committed:
Code review suggestions

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/kubernetes-oidc-caching-validator

timothyb89 committed:
Merge branch 'master' into timothyb89/kubernetes-oidc-caching-validator
